plugins/inputs/aliyuncms/aliyuncms.go

```diff
@@ -0,0 +1,597 @@
+		if err != nil {
+			return errors.Errorf("failed to query metricName list: %v", err)
+		} else if resp.Code != "200" {
+			return errors.Errorf("failed to query metricName list: %v", resp.Message)
+		}
+
+		var datapoints []map[string]interface{}
+		if err = json.Unmarshal([]byte(resp.Datapoints), &datapoints); err != nil {
+			return errors.Errorf("failed to decode response datapoints: %v", err)
+		}
+
+		if len(datapoints) == 0 {
+			lg.logW("No metrics returned from CMS, response msg: %s", resp.Message)
+			break
+		}
+
+	NextDataPoint:
+		for _, datapoint := range datapoints {
+			fields := map[string]interface{}{}
+			datapointTime := int64(0)
+			tags := map[string]string{}
+			for key, value := range datapoint {
+				switch key {
+				case "instanceId", "BucketName":
+					tags[key] = value.(string)
+					if metric.discoveryTags != nil { //discovery can be not activated
+
+						//Skipping data point if discovery data not exist
+						if _, ok := metric.discoveryTags[value.(string)]; !ok &&
+							!metric.AllowDataPointWODiscoveryData {
+							lg.logW("Instance %q is not found in discovery, skipping monitoring datapoint...", value.(string))
+							continue NextDataPoint
+						}
+
+						for k, v := range metric.discoveryTags[value.(string)] {
+							tags[k] = v
+						}
+					}
+				case "userId":
+					tags[key] = value.(string)
+				case "timestamp":
+					datapointTime = int64(value.(float64)) / 1000
+				default:
+					fields[formatField(metricName, key)] = value
+				}
+			}
+			//lg.logW("Datapoint time: %s, now: %s", time.Unix(datapointTime, 0).Format(time.RFC3339), time.Now().Format(time.RFC3339))
+			acc.AddFields(s.measurement, fields, tags, time.Unix(datapointTime, 0))
+		}
+
+		req.NextToken = resp.NextToken
+		more = req.NextToken != ""
+	}
+
+	return nil
+}
+
+//Tag helper
+func parseTag(tagSpec string, data interface{}) (string, string, error) {
+
+	tagKey := tagSpec
+	queryPath := tagSpec
+
+	//Split query path to tagKey and query path
+	if splitted := strings.Split(tagSpec, ":"); len(splitted) == 2 {
+		tagKey = splitted[0]
+		queryPath = splitted[1]
+	}
+
+	tagRawValue, err := jmespath.Search(queryPath, data)
+	if err != nil {
+		return "", "", errors.Errorf("Can't query data from discovery data using query path %q: %v",
+			queryPath, err)
+	}
+
+	if tagRawValue == nil { //Nothing found
+		return "", "", nil
+	}
+
+	tagValue, ok := tagRawValue.(string)
+	if !ok {
+		return "", "", errors.Errorf("Tag value %v parsed by query %q is not a string value",
+			tagRawValue, queryPath)
+	}
+
+	return tagKey, tagValue, nil
+}
+
+func (s *AliyunCMS) prepareTagsAndDimensions(metric *Metric) {
+	var (
+		newData    bool
+		defaulTags = []string{"RegionId:RegionId"}
+	)
+
+	if s.dt == nil { //Discovery is not activated
+		return
+	}
+
+	//Reading all data from buffered channel
+L:
+	for {
+		select {
+		case s.discoveryData = <-s.dt.dataChan:
+			newData = true
+			continue
+		default:
+			break L
+		}
+	}
```

Comment:
doesn't look like the label is required here?

Reply:
The outer `for` cycle is running forever and we need to break it once all buffered data read, otherwise we never go out from the `for` cycle, as `break` without label breaks from inner `select` statement not outer  `for` loop. So I believe the label is needed here.

Reply:
I think @ssoroka was hinting on using a different method instead of endless `for` and a label.

Reply:
@Hipska, thanks for the hint. I oversaw the point.
Maybe smth. like this would be ok:
```go
var moreData = true
for ;moreData;{
select {
  case s.discoveryData = <-s.dt.dataChan:
	newData = true
	continue
  default:
	moreData = false
    }
}
```
What do you think?

Reply:
I have no opinion on this, we'll wait on @ssoroka's comments.

Reply:
@ssoroka You resolved this conversation, but didn't give feedback? Do you mean it is okay to leave like this or what?